Disable soap encoded XmlSerialization tests on desktop

safern opened the pull request:
System.Runtime.Serialization.Xml src and ref contracts are built against netfx not netstandard so the test need to be built against netfx as well to find its correct dependencies in runtime. Now since we are building them against netstandard and run them on netfx it will not find System dependencies as it doesn't have a reference to netstandard. So we should build them against netfx so that it loads it dependencies from mscorlib as it is the ones included from src and ref. 

Fixes: https://github.com/dotnet/corefx/issues/18660

Tracking issue: #18964 

cc: @weshaggard @danmosemsft

shmao commented:
@safern Thanks for the PR. I've assigned #18660 to you. 

The failed tests mentioned in #18660 , however, are the tests under `System.Private.Xml\tests\XmlSerailizer`, not under `System.Runtime.Serialization.Xml`. Do we want to change the props file in the former folder?

Besides, there's `ReflectionOnly` folder in `System.Private.Xml\tests\XmlSerailizer` and it also contains config props file. Do we want to fix it as well?

safern commented:
You're right, I changed the wrong configuration, either way we still want to keep this netfx configuration as the ref and src contracts are built against netfx and not netstandard.

Will update the PR

safern commented:
I updated the PR to what I intended to do from the beginning. Thanks @shmao for pointing that out. Now this do fix: #18660 

I removed the netfx config from Runtime.Serialization.Xml.Tests as I just confirmed it is not needed.

weshaggard commented:
>Now since we are building them against netstandard and run them on netfx it will not find System dependencies as it doesn't have a reference to netstandard. 

These API's are part of netstandard so you should be able to build the tests against netstandard and run on netfx. Blindly adding the netfx configuration isn't necessarily the right fix for this issue.

weshaggard commented:
If there is some fundamental problem with the xml serializer here around netstandard then we need to better understand it to make sure there isn't some blocking issue with it for netstandard libraries.

weshaggard commented:
```
System.InvalidOperationException : Unable to generate a temporary class (result=1).
        error CS0012: The type 'System.Enum' is defined in an assembly that is not referenced. 
       You must add a reference to assembly 'netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'.
```

@shmao this to me seems to indicate that the XmlSerializer is not passing a reference to netstandard.dll whe doing its compiling. Is there something hard-coded in the .NET Framework that adds the references or is it extensible such that we should pass that reference to the serializer in these tests?

safern commented:
So maybe what has the wrong configurations/references is the implementation itself. Will investigate a little deeper to see what is the right fix then.

weshaggard commented:
>So maybe what has the wrong configurations/references is the implementation itself. Will investigate a little deeper to see what is the right fix then.

I don't think it is the implementation as it should have a netfx configuration. There isn't anything wrong with having that.

shmao commented:
@weshaggard We recently enabled XmlSerializer to support soap encoded messages using reflection. The failing tests are all soap encoded serialization related. On Desktop, XmlSerializer uses CodeGen to support soap encoded serialization. I think that's the reason the tests are failing on Desktop.

cc: @zhenlan

safern commented:
> I don't think it is the implementation as it should have a netfx configuration. There isn't anything wrong with having that.

I just verified and System.Private.Xml only has netcoreapp and uap configuration so we should add netstandard configuration.

weshaggard commented:
>I just verified and System.Private.Xml only has netcoreapp and uap configuration so we should add netstandard configuration.

No System.Private.Xml should not have a netstandard configutation for it's library. It is part of netstandard and thus cannot build against it.

weshaggard commented:
>@weshaggard We recently enabled XmlSerializer to support soap encoded messages using reflection. The failing tests are all soap encoded serialization related. On Desktop, XmlSerializer uses CodeGen to support soap encoded serialization. I think that's the reason the tests are failing on Desktop.

If that is the case then we should consider partitioning those tests in someway as opposed to moving everything to a netfx configuration. We eventually plan to run all netstandard configurations on all platforms as well, even if there is a platform specific configuration, so in those runs these tests will still fail. Therefore we need to correctly partition them.

safern commented:
> No System.Private.Xml should not have a netstandard configutation for it's library. It is part of netstandard and thus cannot build against it.

I gotcha. Ok, how should we partition the tests then because in order to include on build time with netfx configuration we should include that config on the test project but we don't want to do that as it is part of the standard?

weshaggard commented:
Given that this is implementation level changes and not API changes you could use the SkipOnFrameworkAttribute for these tests. @shmao @zhenlan I assume the plan is to bring these changes to .NET Framework as well and if so the SkipOnFramework is reasonable, but if not then we should think hard about whether or not we want this behavior difference between the 2 frameworks.

shmao commented:
>  Is there something hard-coded in the .NET Framework that adds the references or is it extensible such that we should pass that reference to the serializer in these tests

@weshaggard I understood the root cause of the issue. When using CodeGen, for all types used by the generated code, XmlSerializer adds the corresponding assembly, `type.Module.Assembly.Location`, as reference assembly. For example, here's one of the test types,

```c#
public enum SoapEncodedTestEnum
{
    A = 0,
    B = 1
}
```
For the generated serializer for the above type, XmlSerializer added the following assemblies as reference assemblies,

```
System.Xml.dll	
mscorlib.dll (for System types, e.g. System.Enum)
System.Xml.XmlSerializer.Tests.dll (for test types)
```

But as System.Xml.XmlSerializer.Tests.dll was built against netstandard, the `System.Enum` type used by `SoapEncodedTestEnum` was from netstandard.dll. The compilation then failed because it couldn't find netstandard.dll.

weshaggard commented:
So what is the correct fix for this then? Why wouldn't it consider the assembly that it is defined in and add the necessary references as well?

shmao commented:
> I assume the plan is to bring these changes to .NET Framework as well and if so the SkipOnFramework is reasonable, but if not then we should think hard about whether or not we want this behavior difference between the 2 frameworks.

Actually no, we don't have plan to bring the changes to desktop. On NetCore we implemented the reflection based serialization to support soap only because we cannot reuse desktop's implementation, which uses CodeDom.

> Given that this is implementation level changes and not API changes you could use the SkipOnFrameworkAttribute for these tests. 

I'd prefer to not skip the tests on netfx. Although the implementations are different on the two platforms, we expect the users to see no difference in behaviors. Running the tests on desktop would help us verifying the serialization behaviors.

safern commented:
> I'd prefer to not skip the tests on netfx. Although the implementations are different on the two platforms, we expect the users to see no difference in behaviors. Running the tests on desktop would help us verifying the serialization behaviors.

So then should we skip the tests and add desktop specific tests in the future? As of now we only care about finding the breaking changes caused by difference in behavior but from what I understand this was an intended change to .Net Core.

shmao commented:
> So what is the correct fix for this then? Why wouldn't it consider the assembly that it is defined in and add the necessary references as well?

I could be wrong, but it seems to me the issue is with the desktop implementation. It cannot see netstandard.dll with the current logic of looking up types. For example, at runtime, the System.Enum being used is from mscorlib.dll. XmlSerializer has no idea it should add netstandard.dll as reference.

weshaggard commented:
Is there a way to pass extra references to it?

shmao commented:
I didn't see any way to pass extra references to XmlSerializer for compiling generated code.

weshaggard commented:
This seems like it will be a real issue then for folks trying to serialize types that are defined in a netstandard library. We need to look into possible solutions for this issue.

safern commented:
I think then we should open an issue marked as 2.0 so that then we can make a decision on how to solve this issue. 

We could either mark the tests as SkipOnTargetFramework and reference them to the new issue or close this PR.

I think the best way to go would be marking them as Skip on desktop and link them to the issue in order to have a clean desktop test run. 

Thoughts?

weshaggard commented:
>I think the best way to go would be marking them as Skip on desktop and link them to the issue in order to have a clean desktop test run.

I'm fine with that but we definitely need to look at the issue for 2.0 because it may become a serialization blocking issue for netstandard libraries which we want to avoid.

shmao commented:
> I'm fine with that but we definitely need to look at the issue for 2.0 because it may become a serialization blocking issue for netstandard libraries which we want to avoid.

Just to make it clear, the issue does not block entire serialization for NetStandard types. 
1. The issue happens only when XmlSerializer use **code gen** to generate serialization assemblies for types. 
2. XmlSerializer uses code gen only for **soap encoded** serializaiton. (Soap encoded serialization was not supported by previous releases of NetCore.)
3. For **non soap encoded** serialization, XmlSerializer uses **IL gen**, which does not have the issue.

safern commented:
Anyway costumers will hit this issue when using soap encoded serialization and building against netstandard and running on netfx. This seems like a blocking issue for soap encoded serialization, because in order to use it then I would need to build against netcoreapp and netfx in order to run the same application in different target groups vs building against want target and then run on different runtimes.

safern commented:
What action would you like to be taken @shmao? I'd go with opening an issue and disabling soap encoded serialization tests linked to that issue.

shmao commented:
@safern I'd prefer to have the soap tests to still run on desktop/netfx so that we know the serialization behaviors for soap encoded messages are compatible between netfx and NetCore. We're still developing the the NetCore way of supporting soap. It's critical to make sure the serialization behavior is compatible with desktop.

weshaggard commented:
>Just to make it clear, the issue does not block entire serialization for NetStandard types.

I agree but it does uncover some broken scenarios which should at least be documented, but hopefully fixed or a workaround for them. 

If we skip tests we should only ship the specific set of test cases that are currently failing an not all of them and file an issue tracking the difference.

safern commented:
@shmao yes the best practice across all of our libraries is to test both netfx and net core and try to be as most compatible as possible, but in this case we just discover that they are not fully compatible, or that at least there is a breaking change when building on netstandard and running on netfx. Still if we add the netfx configuration when running the tests against all configurations it will be built against netstandard and will fail, and we can't just add the netfx configuration and include those tests when targetgroup != netstandard because they will not run when we run the tests for all configurations.

So I think we should skip them (only soap serialization tests) and track an issue for this to think an alternative solution on this problem because adding the netfx configuration is just a patch to have this tests running.

shmao commented:
@safern Sounds good :) Let's skip the failing tests for now then.

safern commented:
Alright, will open the issue and I'll update this PR.

safern commented:
Updated the PR to disable tests and link them to tracking issue.
